src/Nethermind/Nethermind.JsonRpc.TraceStore/DbPersistingBlockTracer.cs

```diff
@@ -0,0 +1,72 @@
+// SPDX-FileCopyrightText: 2022 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Db;
+using Nethermind.Evm.Tracing;
+using Nethermind.Int256;
+using Nethermind.Logging;
+
+namespace Nethermind.JsonRpc.TraceStore;
+
+/// <summary>
+/// Tracer that can store traces of decorated tracer in database
+/// </summary>
+/// <typeparam name="TTrace">Trace type</typeparam>
+/// <typeparam name="TTracer">Transaction tracer type</typeparam>
+public class DbPersistingBlockTracer<TTrace, TTracer> : IBlockTracer where TTracer : class, ITxTracer
+{
+    private readonly IDb _db;
+    private readonly Func<IReadOnlyCollection<TTrace>, byte[]> _serialization;
+    private readonly IBlockTracer _blockTracer;
+    private readonly BlockTracerBase<TTrace, TTracer> _tracerWithResults;
+    private Keccak _currentBlockHash = null!;
+    private long _currentBlockNumber;
+    private readonly ILogger _logger;
+
+    /// <summary>
+    /// Creates the tracer
+    /// </summary>
+    /// <param name="blockTracer">Internal, actual tracer that does the tracing</param>
+    /// <param name="db">Database</param>
+    /// <param name="serialization">Method for serialization</param>
+    /// <param name="logManager"></param>
+    public DbPersistingBlockTracer(BlockTracerBase<TTrace, TTracer> blockTracer,
+        IDb db,
+        Func<IReadOnlyCollection<TTrace>, byte[]> serialization,
```

Comment:
QQ: Why not pass the serializer directly?

Reply:
Old design choice, fixed.